### PR TITLE
Rework tag entity

### PR DESCRIPTION
**Pull request by ChernykhVyacheslav**
Reworked Tag entity so it has relation to specific board.
Now each board has their own list of Tags.

https://trello.com/c/hi1Tbbjx/153-rework-tag-entity

### CODE REVIEW

File: service/src/main/java/com/softserve/borda/services/impl/BoardServiceImpl.java [hunk 1]
```diff
@@ -70,4 +74,17 @@ public boolean deleteBoardById(Long id) {
...
+    public Board addTagToBoard(Long boardId, Long tagId) {
+        Board board = getBoardById(boardId);
+        Tag tag = tagService.getTagById(tagId);
+        board.getTags().add(tag);
+        return boardRepository.save(board);
+    }
+
+    @Override
+    public List<Tag> getAllTagsByBoardId(Long boardId) {
+        return getBoardById(boardId).getTags();
```
Comment (achehrynets) [inline service/src/main/java/com/softserve/borda/services/impl/BoardServiceImpl.java]: Do we really need all data about a board in this case?
I mean, possibly it is better to fetch data about tags from TagRepository?

File: service/src/main/java/com/softserve/borda/services/impl/BoardServiceImpl.java [hunk 2]
```diff
@@ -70,4 +74,17 @@ public boolean deleteBoardById(Long id) {
                 .map(UserBoardRelation::getBoard)
                 .collect(Collectors.toList());
     }
+
+    @Override
+    public Board addTagToBoard(Long boardId, Long tagId) {
+        Board board = getBoardById(boardId);
+        Tag tag = tagService.getTagById(tagId);
+        board.getTags().add(tag);
```
Comment (achehrynets) [inline service/src/main/java/com/softserve/borda/services/impl/BoardServiceImpl.java]: We do not need to touch board to save a tag.
As I understand, Tag entity has boarId inside and you can save only tag without updating the whole board.

**Review by achehrynets — CHANGES_REQUESTED** [review]
Left comments.

File: service/src/main/java/com/softserve/borda/services/impl/BoardServiceImpl.java (hunk 1 above)
Comment (ChernykhVyacheslav) [reply service/src/main/java/com/softserve/borda/services/impl/BoardServiceImpl.java]: Right now we have unidirectional mapping between Board and Tag. Board has list of Tags, but Tags have no link to Board. So I am not sure how I should do it then. Is it just better to use bidirectional mapping? Then I will be able to rework it from TagRepository.

File: service/src/main/java/com/softserve/borda/services/impl/BoardServiceImpl.java (hunk 2 above)
Comment (ChernykhVyacheslav) [reply service/src/main/java/com/softserve/borda/services/impl/BoardServiceImpl.java]: Board has list of tags. Like labels from Trello. And we need them to always be connected to some board. Tag has no board inside right now. This method assumes that Tag has already been created and it just adds it to the board.

**Review by jules-n — APPROVED** [review]
Seems good

File: controller/src/main/java/com/softserve/borda/controllers/BoardController.java [hunk 3]
```diff
@@ -349,35 +350,56 @@ public Board getBoard(@PathVariable Long id) {
...
+        return new ResponseEntity<>(tagDTOs, HttpStatus.OK);
     }
 
     @PutMapping("/boards/{boardId}/tags/{tagId}")
     public ResponseEntity<TagDTO> updateTagByTagId
             (@PathVariable Long tagId,
              @RequestBody TagDTO tagDTO,
-             @PathVariable String boardId) {
+             @PathVariable Long boardId) {
+        if (Objects.isNull(tagDTO.getBoardId())) {
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/BoardController.java]: I think, it is redundant check.
And in general, it better to create `UpdateTagDTO` that contains properties that are available for update and then set `boardId` what you get from path variable.

File: domain/src/main/java/com/softserve/borda/dto/TagDTO.java [hunk 4]
```diff
@@ -8,4 +8,5 @@
     private Long id;
     private String text;
     private String color;
+    private Long boardId;
```
Comment (achehrynets) [inline domain/src/main/java/com/softserve/borda/dto/TagDTO.java:11]: As I understand, we can add a tag to a ticket also, why ticket id is missed here?

Comment (achehrynets) [reply domain/src/main/java/com/softserve/borda/dto/TagDTO.java:11]: Do we have a relation between ticket and tag? - https://github.com/nelfan/BORDA/blob/39db6bea7c22da5dbd236480414b78eb44df15bf/service/src/main/java/com/softserve/borda/services/impl/TicketServiceImpl.java#L86

**Review by achehrynets — COMMENTED** [review]
Left questions for clarification.

File: controller/src/main/java/com/softserve/borda/controllers/BoardController.java (hunk 3 above)
Comment (ChernykhVyacheslav) [reply controller/src/main/java/com/softserve/borda/controllers/BoardController.java]: Will change, thanks for advice

**Pull request merged**